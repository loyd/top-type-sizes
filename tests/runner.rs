use top_type_sizes::*;

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            static SAMPLE: &'static str =
                include_str!(concat!("samples/", stringify!($name), ".txt"));

            let refined_sample = reader::read(SAMPLE.as_bytes()).unwrap();
            let types = parser::parse(&refined_sample).unwrap();
            insta::assert_debug_snapshot!(format!("{}_types", stringify!($name)), types);

            // TODO: check the transformer.

            let output = formatter::format(types, &Default::default());
            println!("{output}");
            // TODO: make snapshots of output.
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
