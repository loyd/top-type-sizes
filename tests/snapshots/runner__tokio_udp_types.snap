---
source: tests/runner.rs
expression: types
---
[
    Type {
        name: "[async fn body@/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.1/src/net/udp.rs:1345:86: 1350:6]",
        size: 200,
        align: 8,
        kind: Enum(
            EnumType {
                discriminant_size: Some(
                    1,
                ),
                variants: [
                    EnumVariant {
                        name: "Unresumed",
                        size: 192,
                        items: [
                            Padding(
                                159,
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "buf",
                                    size: 16,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Padding(
                                8,
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "self",
                                    size: 8,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Returned",
                        size: 192,
                        items: [
                            Padding(
                                159,
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "buf",
                                    size: 16,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Padding(
                                8,
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "self",
                                    size: 8,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Panicked",
                        size: 192,
                        items: [
                            Padding(
                                159,
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "buf",
                                    size: 16,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Padding(
                                8,
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "self",
                                    size: 8,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Suspend0",
                        size: 184,
                        items: [
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "buf",
                                    size: 16,
                                    align: Some(
                                        8,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "__awaitee",
                                    size: 144,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "buf",
                                    size: 16,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "self",
                                    size: 8,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "self",
                                    size: 8,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                ],
            },
        ),
        end_padding: Some(
            7,
        ),
    },
]
